File: cpp/src/mav_deepControl.cpp
/*****************************************************************************
-----------PRATYUSH VARSHNEY-----------------
------------pratyushvarshney@cse.iitk.ac.in-------------
*****************************************************************************/

#include <ros/ros.h>
#include <sensor_msgs/Imu.h>
#include <nav_msgs/Odometry.h>
#include <mav_msgs/RollPitchYawrateThrust.h>
#include <geometry_msgs/PoseStamped.h>
#include <geometry_msgs/Quaternion.h>
#include "include/mav_nonlinear_mpc/params.hpp"
// #include <params.hpp>
using namespace Eigen;
// using namespace Eigen;
#define NO_OF_STATES 6

geometry_msgs::PoseStamped pose_;
mav_msgs::RollPitchYawrateThrust mpc_;
nav_msgs::Odometry odom_;

void pose_cb(const geometry_msgs::PoseStamped msg)
{
    pose_ = msg;
    return;
}

void mpc_cb(const mav_msgs::RollPitchYawrateThrust msg)
{
    mpc_ = msg;
    return;
}

void odom_cb(const nav_msgs::Odometry msg)
{
    odom_ = msg;
    return;
}



static void toEulerAngle(const geometry_msgs::Quaternion& q, double& roll, double& pitch, double& yaw)
{
  // roll (x-axis rotation)
  double sinr_cosp = +2.0 * (q.w * q.x + q.y * q.z);
  double cosr_cosp = +1.0 - 2.0 * (q.x * q.x + q.y * q.y);
  roll = atan2(sinr_cosp, cosr_cosp);

  // pitch (y-axis rotation)
  double sinp = +2.0 * (q.w * q.y - q.z * q.x);
  if (fabs(sinp) >= 1)
    pitch = copysign(M_PI / 2, sinp); // use 90 degrees if out of range
  else
    pitch = asin(sinp);

  // yaw (z-axis rotation)
  double siny_cosp = +2.0 * (q.w * q.z + q.x * q.y);
  double cosy_cosp = +1.0 - 2.0 * (q.y * q.y + q.z * q.z);  
  yaw = atan2(siny_cosp, cosy_cosp);
}

float max(float a, float b){
  if(a < b)
    return b;
  else
    return a;
}

int main(int argc, char** argv)
{
  ros::init(argc, argv, "weight_mpc");
  ros::NodeHandle nh;

  ros::Subscriber pose_sub = nh.subscribe<geometry_msgs::PoseStamped>("command/pose",1000,pose_cb);
  ros::Subscriber odom_sub = nh.subscribe<nav_msgs::Odometry>("ground_truth/odometry",1000,odom_cb);
  ros::Subscriber mpc_sub = nh.subscribe<mav_msgs::RollPitchYawrateThrust>("rpyth",1000,mpc_cb);
  
  ros::Publisher  rpyth_pub = nh.advertise<mav_msgs::RollPitchYawrateThrust>("command/roll_pitch_yawrate_thrust", 1000);
  
  ros::Rate loop_rate(100);
  weights np;

  double x,y,z, vx,vy,vz, rs,ps,ys;
  mav_msgs::RollPitchYawrateThrust rpyth;



  while (nh.ok()) 
  {
    double roll = 0.0, pitch = 0.0, yaw = 0.0;
    toEulerAngle(odom_.pose.pose.orientation, roll, pitch, yaw);

    x = odom_.pose.pose.position.x;
    y = odom_.pose.pose.position.y;
    z = odom_.pose.pose.position.z;


    vx = odom_.twist.twist.linear.x;
    vy = odom_.twist.twist.linear.y;
    vz = odom_.twist.twist.linear.z;

    VectorXf target(NO_OF_STATES);
    target(0) = pose_.pose.position.x;
    target(1) = pose_.pose.position.y;
    target(2) = 0.0;

    target(3) = 0.0;
    target(4) = 0.0;
    target(5) = 0.0;

    VectorXf state_(NO_OF_STATES);
    
    state_(0) = x;
    state_(1) = y;

    state_(2) = vx;
    state_(3) = vy;

    state_(4) = roll;
    state_(5) = pitch;




    Matrix<float,1,NO_OF_STATES> input_;

    for(int i = 0; i < NO_OF_STATES; i++)
      input_(0,i) = state_(i) -target(i);

    
    for(int i = 0; i < NO_OF_STATES; i++){ 
      input_(0,i) = (input_(0,i) - np.mean(i,0)) / np.std(i,0);
    }

    

    VectorXf controls_(2);

    // LAYER 1
    Matrix<float,1,64> op1_w = (input_ * np.layer_1_weight.transpose());
    Matrix<float,1,64> op1 = op1_w + np.bias_1.transpose();

    for(int i = 0; i < 64; i++)
      op1(0,i) = max(0, op1(0,i));


    //LAYER 2

    Matrix<float,1,64> op2_w = (op1 * np.layer_2_weight.transpose());
    Matrix<float,1,64> op2 = op2_w + np.bias_2.transpose();

    for(int i = 0; i < 64; i++)
      op2(0,i) = max(0, op2(0,i));

    


    // LAYER 3
    Matrix<float,1,2> op3_w = (op2 * np.layer_3_weight.transpose());
    Matrix<float,1,2> op3_b = op3_w + np.bias_3.transpose();
    ArrayWrapper<Matrix<float,1,2> > op3(op3_b); 
    op3.tanh();

    rpyth.header.stamp = ros::Time::now();
    rpyth.roll  = op3(0,0);
    rpyth.pitch = op3(0,1);
    rpyth.yaw_rate = mpc_.yaw_rate;
    rpyth.thrust.z = mpc_.thrust.z;

    rpyth_pub.publish(rpyth);
    ros::spinOnce(); 

    loop_rate.sleep();
  }
  return 0;
}

File: python/mav_deepControl.py
#!/usr/bin/env python


import math
import numpy as np
import rospy
from mav_msgs.msg import Actuators
from nav_msgs.msg import Odometry
from mav_msgs.msg import RollPitchYawrateThrust
from geometry_msgs.msg import PoseStamped, TwistStamped
from sensor_msgs.msg import Imu
from trajectory_msgs.msg import MultiDOFJointTrajectory
import argparse

from std_msgs.msg import String

PI = 3.14

weights_dir = './wt/'
output_dir = './wt/'

weights_1 = np.load(weights_dir + 'weights_1.npy').transpose()
bias_1   = np.load(weights_dir + 'bias_1.npy').transpose()


weights_2 = np.load(weights_dir + 'weights_2.npy').transpose()
bias_2  = np.load(weights_dir + 'bias_2.npy').transpose()


weights_3 = np.load(weights_dir + 'weights_3.npy').transpose()
bias_3   = np.load(weights_dir + 'bias_3.npy').transpose()

mean = np.load(output_dir + 'mean.npy')
std  = np.load(output_dir + 'std.npy')

mean_x = np.load(output_dir_x + 'mean.npy')
std_x  = np.load(output_dir_x + 'std.npy')





# print('mean', mean)
# print('std', std)
# exit()

target_flag_pub  = False

vrpn = TwistStamped()
def vrpn_cb(data):
	global vrpn
	vrpn = data


target_pose = MultiDOFJointTrajectory()
def target_cb(data):
	global target_pose
	target_pose = data
	global target_flag_pub
	target_flag_pub = True

mpc_u = RollPitchYawrateThrust()
def mpc_cb(data):
	global mpc_u
	mpc_u = data


state_machine = String()
def state_cb(data):
	global state_machine
	state_machine = data


def quaternion_to_euler_angle(w, x, y, z):
	ysqr = y * y

	t0 = +2.0 * (w * x + y * z)
	t1 = +1.0 - 2.0 * (x * x + ysqr)
	X = math.degrees(math.atan2(t0, t1))

	t2 = +2.0 * (w * y - z * x)
	t2 = +1.0 if t2 > +1.0 else t2
	t2 = -1.0 if t2 < -1.0 else t2
	Y = math.degrees(math.asin(t2))

	t3 = +2.0 * (w * z + x * y)
	t4 = +1.0 - 2.0 * (ysqr + z * z)
	Z = math.degrees(math.atan2(t3, t4))

	return X, Y, Z


odom = Odometry()
def odom_cb(data):
	global odom
	odom = data


def do_control( ):
	
	odom_sub = rospy.Subscriber('/f_450/odometry',Odometry, odom_cb, queue_size=100)
	vrpn_sub = rospy.Subscriber('/f_450/vrpn_client_node/f_450/twist',TwistStamped,vrpn_cb, queue_size=100)
	state_sub = rospy.Subscriber('/f_450/state_machine/state_info', String, state_cb, queue_size=100)
	target_pose_sub = rospy.Subscriber('/f_450/command/current_reference', MultiDOFJointTrajectory ,target_cb, queue_size=100)
	mpc_sub = rospy.Subscriber('/f_450/command/roll_pitch_yawrate_thrust',RollPitchYawrateThrust, mpc_cb, queue_size=100)

	control_pub = rospy.Publisher("/f_450/mavros/setpoint_raw/roll_pitch_yawrate_thrust", RollPitchYawrateThrust,queue_size=100)

	rospy.init_node('controller',anonymous=True)
	rate =  rospy.Rate(50.0)
	global target_flag_pub	

	# intial Target
	target = np.array([0.0,0.0,0.5, 0.,0.,0., 0.,0.],ndmin=2)
	

	while not rospy.is_shutdown():

		x_f = odom.pose.pose.position.x
		y_f = odom.pose.pose.position.y
		z_f = odom.pose.pose.position.z

		vx_f = odom.twist.twist.linear.x
		vy_f = odom.twist.twist.linear.y
		vz_f = odom.twist.twist.linear.z

		(roll,pitch, yaw) = quaternion_to_euler_angle(odom.pose.pose.orientation.w, odom.pose.pose.orientation.x , odom.pose.pose.orientation.y, odom.pose.pose.orientation.z)
		r_f = math.radians(roll)
		p_f = math.radians(pitch)
		yaw_f = math.radians(yaw)

		rs_f = float(vrpn.twist.angular.x)
		ps_f = float(vrpn.twist.angular.y)
		ys_f = float(vrpn.twist.angular.z)

		if target_flag_pub == True:
			target = np.array([target_pose.points[0].transforms[0].translation.x , target_pose.points[0].transforms[0].translation.y,target_pose.points[0].transforms[0].translation.z, 0.,0.,0., 0.,0.],ndmin=2)
		
		

		target_flag_pub = False


		state = np.array([x_f,y_f,z_f, vx_f,vy_f,vz_f, r_f,p_f],ndmin=2)

		inputs = state - target
		inputs_x = (inputs-mean_x)/std_x
		inputs = (inputs-mean)/std

		print(str(state_machine.data))

		if str(state_machine.data) == 'PositionHold':
			
			print('MPC: ')
			controls = np.ravel(get_actions(inputs))
			controls_x = np.ravel(get_actions_x(inputs_x))
			rpyth_mpc = RollPitchYawrateThrust()

			rpyth_mpc.header.stamp = rospy.Time.now()	
			rpyth_mpc.roll = controls[0]
			rpyth_mpc.pitch = controls_x[1]
			rpyth_mpc.yaw_rate = mpc_u.yaw_rate
			rpyth_mpc.thrust.z = (controls[2] * 7.5) + 7.5	
			control_pub.publish(rpyth_mpc)
			

		elif str(state_machine.data) == 'RcTeleOp' or str(state_machine.data) == 'RemoteControl' or str(state_machine.data) == 'HaveOdometry' :
			print('MPC')
			rpyth_mpc = RollPitchYawrateThrust()

			rpyth_mpc.header.stamp = rospy.Time.now()	
			rpyth_mpc.roll = mpc_u.roll
			rpyth_mpc.pitch = mpc_u.pitch
			rpyth_mpc.yaw_rate = mpc_u.yaw_rate
			rpyth_mpc.thrust.z = mpc_u.thrust.z
			
			
			control_pub.publish(rpyth_mpc)


		rate.sleep()



def get_actions(inputs):

	x = np.matmul(inputs, weights_1)
	x = np.add(x, bias_1)
	x = np.maximum(x, 0, x)


	x = np.matmul(x, weights_2)
	x = np.add(x, bias_2)
	x = np.maximum(x, 0, x)

	x = np.matmul(x, weights_3)
	x = np.add(x, bias_3)
	controls = np.tanh(x)	
		

	return controls
	


if __name__ == '__main__':
	try:
		do_control()

	except rospy.ROSInterruptException:
		pass
